pkg/ipcache/ipcache.go

```diff
@@ -180,77 +189,193 @@ func allowOverwrite(existing, new Source) bool {
+	scopedLog := log.WithFields(logrus.Fields{
+		logfields.IPAddr:   ip,
+		logfields.Identity: newIdentity,
+	})
+
 	ipc.mutex.Lock()
 	defer ipc.mutex.Unlock()
 
-	if existingIdentity, found := ipc.ipToIdentityCache[IP]; found {
-		if !allowOverwrite(existingIdentity.Source, identity.Source) {
+	var cidr *net.IPNet
+	var oldIdentity *identity.NumericIdentity
+	callbackListeners := true
+
+	cachedIdentity, found := ipc.ipToIdentityCache[ip]
+	if found {
+		if !allowOverwrite(cachedIdentity.Source, newIdentity.Source) {
 			return false
 		}
 
-		// Skip update if IP is already mapped to the given identity
-		if existingIdentity == identity {
+		// Skip update if IP is already mapped to the given identity.
+		// Ignore the sources of those identities.
+		if cachedIdentity.ID == newIdentity.ID {
 			return false
 		}
-	}
-
-	// An update is treated as a deletion and then an insert.
-	ipc.deleteLocked(IP)
-
-	log.WithFields(logrus.Fields{
-		logfields.IPAddr:   IP,
-		logfields.Identity: identity,
-	}).Debug("Upserting into ipcache layer")
 
-	// Update both maps.
-	ipc.ipToIdentityCache[IP] = identity
-
-	_, found := ipc.identityToIPCache[identity.ID]
-	if !found {
-		ipc.identityToIPCache[identity.ID] = map[string]struct{}{}
+		oldIdentity = &cachedIdentity.ID
 	}
-	ipc.identityToIPCache[identity.ID][IP] = struct{}{}
 
-	// Add a reference for the prefix length if this is a CIDR.
-	if _, cidr, err := net.ParseCIDR(IP); err == nil {
+	// Endpoint IP identities take precedence over CIDR identities, so if the
+	// IP is a full CIDR prefix and there's an existing equivalent endpoint IP,
+	// don't notify the listeners.
+	var err error
+	if _, cidr, err = net.ParseCIDR(ip); err == nil {
+		// Add a reference for the prefix length if this is a CIDR.
 		pl, bits := cidr.Mask.Size()
 		switch bits {
 		case net.IPv6len * 8:
 			refPrefixLength(ipc.v6PrefixLengths, pl)
 		case net.IPv4len * 8:
 			refPrefixLength(ipc.v4PrefixLengths, pl)
 		}
+
+		ones, bits := cidr.Mask.Size()
+		if ones == bits {
+			if _, endpointIPFound := ipc.ipToIdentityCache[cidr.IP.String()]; endpointIPFound {
+				scopedLog.Info("Upserting CIDR shadowed by endpoint IP")
+				// Skip calling back the listeners, since the endpoint IP has
+				// precedence over the new CIDR.
+				callbackListeners = false
+			}
+		}
+	} else if endpointIP := net.ParseIP(ip); endpointIP != nil { // Endpoint IP.
+		// Convert the endpoint IP into an equivalent full CIDR.
+		bits := net.IPv6len * 8
+		if endpointIP.To4() != nil {
+			bits = net.IPv4len * 8
+		}
+		cidr = &net.IPNet{
+			IP:   endpointIP,
+			Mask: net.CIDRMask(bits, bits),
+		}
+
+		// Check whether the upserted endpoint IP will shadow that CIDR, and
+		// replace its mapping with the listeners if that was the case.
+		if !found {
+			if cidrIdentity, cidrFound := ipc.ipToIdentityCache[cidr.String()]; cidrFound {
+				if cidrIdentity.ID != newIdentity.ID {
+					scopedLog.Info("Upserting endpoint IP shadowing CIDR")
```

Comment:
`New endpoint IP started shadowing existing CIDR to identity mapping`

Reply:
Done.